### ISSUE TITLE
Generalize common R functions between SMC and BM

### ISSUE THREAD

**Issue by wleoncio** [ocbe-uio/BayesMallows#80]
# Problem

We need a better solution for how SMC uses modified version of two functions already on BM:

1. `compute_posterior_intervals()`
2. `compute_consensus()`

PR #78 works around this issue by appending the SMC versions of these functions with `_smc`, but since they are essentially the same functions, they should be better connected. 

# Solution

These are my assumptions:

- Workflow-wise, the departure point between BM and SMC is in the `smc_mallows_new_users_complete()` function.
- The relevant BM functions assume the input to be of class `BayesMallows`

The plan is as follows:

- [x] Modify `smc_mallows_new_users_complete()` so that its output is not a simple list, but an object of (a new) class `SMCMallows` (or some other suitable name).
- [x] Append `compute_posterior_intervals()` and `compute_consensus()` with `.BayesMallows()` for S3 method dispatching.
- [x] Do the same on the internal subfunctions of the aforementioned functions, such as `.compute_cp_consensus()`. This is where things get more complicated, as these functions seem to use a simple data frame as input. Gotta think a little about the solution with the fewest changes to the original code.
- [x] Create `compute_posterior_intervals()` and `compute_consensus()` generic S3 functions.
- [x] Make sure everything still works (unit tests are passing).
- [x] Rename the derivative BayesMallows (the ones created by the new generics) to `.BM` or something else
- [x] Rename the `*_smc()` functions as methods
- [x] Consider moving the functions from the step above so generics and their respective methods live in the same file
- [x] Test and check package
- [x] Open PR for merging into `smc-mallows`

@osorensen, can I get an "Amen!"? 😆  I've only considered the plan looking at the examples for `compute_posterior_intervals()` and `compute_consensus()`, but I hope it looks fine considering the overall usage of the package as well. In any case, I won't have time to work on this today, so I am also writing this to bring to the meeting tomorrow.

**Comment by osorensen** [comment]
@wleoncio , you have my Amen! I'm not into the details here, but as long as the original BayesMallows unit tests pass, I have full trust in you :)

**Comment by wleoncio** [comment]
Currently working on this on a local feature branch. For the record (as mentioned in the meeting with @anjastein and @osorensen):

- `compute_posterior_intervals()` is working as a generic function with two methods: one for `BayesMallows` objects another for `SMCMallows` objects;
- `compute_posterior_intervals.BayesMallows()` works fine;
- `compute_posterior_intervals.SMCMallows()` works fine when called with the output of `smc_mallows_new_users_complete()`, which is a length-2 list of class `SMCMallows`;
- `compute_posterior_intervals.SMCMallows()` _does not work_ with the output of `smc_processing()`, which is a data frame. This situation is happens when  using the `compute_posterior_intervals_rho()` function (see code [here](https://github.com/ocbe-uio/BayesMallows/blob/824e37b68d9b6dc46256704cb679a9485c1739d9/R/smc_post_processing_functions.R#L412-L419)).

So my plan is to do some digging into the SMC functions above to find a way to either:

1. standardize the output of `smc_mallows_new_users_complete()` and `smc_processing()`; or
2. Make `smc_processing()` output a different method (e.g. `SMCMallows-dataframe`) and create a method for it on `compute_posterior_intervals()`

The first solution sounds cleaner, but takes a bit more work; the second one is simpler, but will create duplicated code, since `compute_posterior_intervals.SMCMallows-dataframe()` would be nearly identical to `compute_posterior_intervals.SMCMallows()`.

In any case, Issue #81 is more important than this, so I'll be focusing my time on that and get back to this one later (or whenever I need a distraction).

**Comment by wleoncio** [comment]
New task: don't disrupt the original structure from the BayesMallows class (i.e., that from the output of `compute_mallows()`. Change the `.BayesMallows` and `.SMCMallows` classes that came later.

See conversation starting on https://github.com/ocbe-uio/BayesMallows/pull/161#pullrequestreview-817632749 for more info.

**Comment by wleoncio** [comment]
Hi @anjastein,

I can't find an example for the `compute_consensus_smc()` function, could you write a short one for the package? This would be a lot helpful in making sure the integration of that to the original `compute_consensus()` goes smoothly.

Thanks,
Waldir

**Comment by anjastein** [comment]
Okay, so `compute_consensus_smc` is called by the function `compute_rho_consensus, so should I make a unit test for this function instead?

**Comment by anjastein** [comment]
I have attached a gist with the proposed unit test. I've noticed that the function does not return a tibble table when `type="MAP", so the function may need a minor change if necessary
https://gist.github.com/anjastein/c05eb5ce5fae290671e92e62565125d7

**Comment by wleoncio** [comment]
> Okay, so `compute_consensus_smc` is called by the function `compute_rho_consensus`, so should I make a unit test for this function instead?

Not necessarily, I'm starting to think there's some confusion on my part.

On the BayesMallows side, there are two internal functions, `.compute_cp_consensus` and `.compute_map_consensus`, which are called by the exported function `compute_consensus`.

The SMC-Mallows side also contains those same functions (currently with an `_smc` suffix). However, `compute_consensus_smc` is not an exported function. As you said, it is called by `cmopute_rho_consensus`, which _is_ exported.

In other words, as far as API-level functions go, the user either runs `compute_consensus()` for BayesMallows or `comptue_rho_consensus()` for SMC-Mallows. Is that correct?

**Comment by wleoncio** [comment]
> I have attached a gist with the proposed unit test. I've noticed that the function does not return a tibble table when `type="MAP", so the function may need a minor change if necessary https://gist.github.com/anjastein/c05eb5ce5fae290671e92e62565125d7


Thanks for this, I've opened issue #163 to handle the case. Extra unit tests are always welcome too, so thanks for that! :)

**Comment by anjastein** [comment]
Yes, however, the functions `.compute_cp_consensus` and `.compute_map_consensus` have been edited in the SMC case - I think it was to do with the fact that I was dealing with 1 cluster and not `n_clusters`. The solution may be to include extra information in the output of the SMC algorithms so that we can just call the original  `.compute_cp_consensus` and `.compute_map_consensus`.

**Comment by wleoncio** [comment]
OK, thank you for the explanation. The plan is to have `.compute_map/cp_consensus` as generic functions, which are able to dispatch (i.e. call) the respective methods (subfunctions) depending on whether its input has a BM or an SMC-M class. So that's why understanding the parent functions is important, because we need to set those classes in the correct places.
